src/state.rs

```diff
@@ -136,6 +136,136 @@ pub fn advertised_node_for_surface(w: &WlSurface, dh: &DisplayHandle) -> Option<
+            delta: 0.0,
+            history: VecDeque::new(),
+        }
+    }
+
+    pub fn update(&mut self, movement: Point<f64, Logical>, timestamp: Duration) -> bool {
+        let first_update = self.direction.is_none();
+
+        if first_update {
+            // Find proper direction
+            self.direction = if movement.x.abs() > movement.y.abs() {
+                if movement.x > 0.0 {
+                    Some(Direction::Right)
+                } else {
+                    Some(Direction::Left)
+                }
+            } else {
+                if movement.y > 0.0 {
+                    Some(Direction::Down)
+                } else {
+                    Some(Direction::Up)
+                }
+            }
+        }
+
+        let delta = match self.direction {
+            Some(Direction::Left) => -movement.x,
+            Some(Direction::Right) => movement.x,
+            Some(Direction::Up) => -movement.y,
+            Some(Direction::Down) => movement.y,
+            None => 0.0,
+        };
+
+        self.push(delta, timestamp);
+        first_update
+    }
+
+    /// Pushes a new reading into the tracker.
+    fn push(&mut self, delta: f64, timestamp: Duration) {
+        // For the events that we care about, timestamps should always increase
+        // monotonically.
+        if let Some(last) = self.history.back() {
+            if timestamp < last.timestamp {
+                trace!(
+                    "ignoring event with timestamp {timestamp:?} earlier than last {:?}",
+                    last.timestamp
+                );
+                return;
+            }
+        }
+
+        self.history.push_back(SwipeEvent { delta, timestamp });
+        self.delta += delta;
+
+        self.trim_history();
+    }
+
+    /// Computes the current gesture velocity.
+    pub fn velocity(&self) -> f64 {
+        let (Some(first), Some(last)) = (self.history.front(), self.history.back()) else {
+            return 0.;
+        };
+
+        let total_time = (last.timestamp - first.timestamp).as_secs_f64();
+        if total_time == 0. {
+            return 0.;
+        }
+
+        let total_delta = self.history.iter().map(|event| event.delta).sum::<f64>();
+        total_delta / total_time
+    }
+
+    /// Computes the gesture end position after decelerating to a halt.
+    pub fn projected_end_pos(&self) -> f64 {
+        let vel = self.velocity();
+        self.delta - vel / (1000. * DECELERATION_TOUCHPAD.ln())
+    }
+
+    fn trim_history(&mut self) {
+        let Some(&SwipeEvent { timestamp, .. }) = self.history.back() else {
+            return;
+        };
+
+        while let Some(first) = self.history.front() {
+            if timestamp <= first.timestamp + HISTORY_LIMIT {
+                break;
+            }
+
+            let _ = self.history.pop_front();
+        }
+    }
+}
+
+impl Default for GestureState {
+    fn default() -> Self {
+        GestureState::new(0)
+    }
+}
+
```

Comment:
I think this code deserves it's own module it this point.

Reply:
done in 26b8cd8ec581870c8e194c08745734dac96d493a